File: Daily/Within_Two_Edits_Of_Dictionary/Within_Two_Edits_Of_Dictionary.cpp
// Within_Two_Edits_Of_Dictionary.cpp
//
// Compile locally with:
// clang++ Within_Two_Edits_Of_Dictionary.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<string> twoEditWords(vector<string> &queries, vector<string> &dictionary)
    {
        vector<string> results;
        int n = queries[0].size();
        int diff = 0;
        for (string q : queries)
        {
            for (string word : dictionary)
            {
                diff = 0;
                for (int index = 0; index < n; index++)
                {
                    if (q[index] != word[index])
                    {
                        diff++;
                        if (diff > 2)
                        {
                            break;
                        }
                    }
                }
                if (diff <= 2)
                {
                    results.push_back(q);
                    break;
                }
            }
        }
        return results;
    }
};

int main()
{
    cout << "Running Within_Two_Edits_Of_Dictionary.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: Daily/Within_Two_Edits_Of_Dictionary/within__two__edits__of__dictionary.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def twoEditWords(self, queries: List[str], dictionary: List[str]) -> List[str]:
        results = []

        for q in queries:
            for word in dictionary:
                diff = 0
                for index in range(len(word)):
                    if q[index] != word[index]:
                        diff += 1
                        if diff > 2:
                            break
                if diff <= 2:
                    results.append(q)
                    break

        return results


if __name__ == "__main__":
    sol = Solution()
    queries = ["word", "note", "ants", "wood"]
    dictionary = ["wood", "joke", "moat"]
    print(sol.twoEditWords(queries, dictionary))
    print("Running Solution...")


"""
            class Solution:
    def twoEditWords(self, queries: List[str], dictionary: List[str]) -> List[str]:
        dset = set(dictionary)

        def check(word, remaining):
            if remaining == 0 or word in dset:
                return word in dset
            n = len(word)
            for i in range(n):
                for dword in dset:
                    tmp = word[:i] + dword[i] + word[i + 1 :]
                    if check(tmp, remaining - 1):
                        return True
            return False

        return [q for q in queries if check(q, 2)]

            """
